### PR TITLE
Add CodeGenWithLocation

### LINKED ISSUE #55: CodeGenWithLocation
By analogy to https://github.com/shapesecurity/shift-java/pull/136.

#3 could be built on top of this in a fairly straightforward way, I think.

### PR DESCRIPTION
**Pull request by bakkot**
Fixes #55.

~This adds a runtime dependency on `shift-spec`, which maybe isn't ideal, but I don't think it's that bad.~ No longer true.

Commits have been organized for ease of reviewing (you know, relatively speaking).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,6 +24,16 @@ import codegen from "shift-codegen";
 let programSource = codegen(/* Shift format AST */);
 ```
 
+Location information is available in environments which support WeakMap via an alternative interface:
+```js
+let tree = parseScript('foo(); bar;');
+
+import codeGenWithLocation from "shift-codegen";
```
Comment (bakkot) [inline README.md]: Doc is wrong here; should fix before merging

**Review by michaelficarra — APPROVED** [review]
I don't really like how much the with-location code gets its hands all up in the inner workings of the codegen, but I don't have an alternative suggestion right now. LGTM.

**Comment by bakkot** [comment]
@michaelficarra The best alternative I have is to make the `CodeRep` objects dumb and put their `emit` code elsewhere, possibly in `TokenStream`. (Hello, expression problem.) That would avoid the intrusive `rep.emit =`, at least.

Doesn't really seem worth it at this stage, though.

**Comment by bakkot** [comment]
Now includes version bump commit.

**Pull request merged**